plugins/funind/indfun.ml

```diff
@@ -37,111 +41,102 @@ let choose_dest_or_ind scheme_info args =
-         begin
-           match EConstr.kind sigma f with
-           | Const (c',u) ->
-              let princ_option =
-                let finfo = (* we first try to find out a graph on f *)
-                  try find_Function_infos c'
-                  with Not_found ->
-                    user_err  (str "Cannot find induction information on "++
-                                       Printer.pr_leconstr_env (Tacmach.pf_env g) sigma (mkConst c') )
-                in
-                match Tacticals.elimination_sort_of_goal g with
-                | InSProp -> finfo.sprop_lemma
-                | InProp -> finfo.prop_lemma
-                | InSet -> finfo.rec_lemma
-                | InType -> finfo.rect_lemma
-              in
-              let princ,g' =  (* then we get the principle *)
-                try
-                  let g',princ =
-                    Tacmach.pf_eapply (Evd.fresh_global) g  (GlobRef.ConstRef (Option.get princ_option )) in
-                  princ,g'
-                with Option.IsNone ->
-                  (*i If there is not default lemma defined then,
-                          we cross our finger and try to find a lemma named f_ind
-                          (or f_rec, f_rect) i*)
-                  let princ_name =
-                    Indrec.make_elimination_ident
-                      (Label.to_id (Constant.label c'))
-                      (Tacticals.elimination_sort_of_goal g)
-                  in
-                  try
-                    let princ_ref = const_of_id princ_name in
-                    let (a,b) = Tacmach.pf_eapply (Evd.fresh_global) g princ_ref in
-                    (b,a)
-                    (* mkConst(const_of_id princ_name ),g (\* FIXME *\) *)
-                  with Not_found -> (* This one is neither defined ! *)
-                    user_err  (str "Cannot find induction principle for "
-                                     ++ Printer.pr_leconstr_env (Tacmach.pf_env g) sigma (mkConst c') )
+    match princl with
+    | None -> (* No principle is given let's find the good one *)
+      begin
+        match EConstr.kind sigma f with
+        | Const (c',u) ->
+          let princ_option =
+            let finfo = (* we first try to find out a graph on f *)
+              try find_Function_infos c'
+              with Not_found ->
+                user_err  (str "Cannot find induction information on "++
+                           Printer.pr_leconstr_env (pf_env gl) sigma (mkConst c') )
+            in
+            match elimination_sort_of_goal gl with
+            | InSProp -> finfo.sprop_lemma
+            | InProp -> finfo.prop_lemma
+            | InSet -> finfo.rec_lemma
+            | InType -> finfo.rect_lemma
+          in
+          let princ =  (* then we get the principle *)
+            try
+              let sigma, princ = Evd.fresh_global (pf_env gl) (project gl) (GlobRef.ConstRef (Option.get princ_option)) in
```

Comment:
Why not simply match on `princ_option`?

Reply:
Good tip, thanks! I think I've made the code a tad cleaner now.